// audic/dist/browser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const ow_1 = __importDefault(require("ow"));
module.exports = class Audic {
    constructor(src) {
        ow_1.default(src, ow_1.default.optional.string);
        this._el = new Audio(src);
    }
    async play() {
        await this._el.play();
    }
    async pause() {
        this._el.pause();
    }
    get volume() {
        return this._el.volume;
    }
    set volume(value) {
        ow_1.default(value, ow_1.default.number.inRange(0, 1));
        this._el.volume = value;
    }
    get duration() {
        return this._el.duration;
    }
    get playing() {
        return !this._el.paused;
    }
    get src() {
        return this._el.src;
    }
    set src(value) {
        ow_1.default(value, ow_1.default.string);
        this._el.src = value;
    }
    get currentTime() {
        return this._el.currentTime;
    }
    set currentTime(value) {
        ow_1.default(value, ow_1.default.number.integer.greaterThanOrEqual(0));
        this._el.currentTime = value;
    }
    destroy() {
        this._el.remove();
    }
};
